src/Servers.jl

```diff
@@ -410,108 +118,110 @@ function handle_connection(f::Function, c::Connection;
+            end
             @error exception=(e, stacktrace(catch_backtrace()))
         end
-        close(t)
-    finally
-        final_transaction && close(t.c)
+        count += 1
     end
     return
 end
 
-"""
-Execute stream processing function `f`.
-If there is an error and the stream is still open,
-send a 500 response with the error message.
+getinet(host::String, port::Integer) = Sockets.InetAddr(parse(IPAddr, host), port)
+getinet(host::IPAddr, port::Integer) = Sockets.InetAddr(host, port)
 
-Close the `Stream` for read and write (in case `f` has not already done so).
-"""
-function handle_stream(f::Function, http::Stream)
+function listen(h::Handler, host::Union{IPAddr, String}, port::Integer;
+    tcpref::Union{Ref, Nothing}=nothing,
+    reuseaddr::Bool=false,
+    sslconfig::Union{MbedTLS.SSLConfig, Nothing}=nothing,
+    tcpisvalid::Union{Function, Nothing}=nothing,
+    ratelimit::Union{Rational{Int}, Nothing}=nothing,
+    connectioncounter::Ref{Int}=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
 
-    try
-        if applicable(f, http)
-            f(http)
-        else
-            handle_request(f, http)
-        end
-    catch e
-        if isopen(http) && !iswritable(http)
-            @error exception=(e, stacktrace(catch_backtrace()))
-            http.message.response.status = 500
-            startwrite(http)
-            write(http, sprint(showerror, e))
+    inet = getinet(host, port)
+    if tcpref !== nothing
+        tcpserver = tcpref[]
+    elseif reuseaddr
+        tcpserver = Sockets.TCPServer(; delay=false)
+        if Sys.islinux() || Sys.isapple()
+            rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Cvoid},), tcpserver.handle)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         else
-            rethrow(e)
+            @warn "reuseaddr=true may not be supported on this platform: $(Sys.KERNEL)"
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         end
+        Sockets.listen(tcpserver)
+    else
+        tcpserver = Sockets.listen(inet)
     end
+    verbose && println("Listening on: $host:$port")
 
-    closeread(http)
-    closewrite(http)
-    return
+    if tcpisvalid === nothing
+        tcpisvalid = ratelimit === nothing ? x->true : x->check_rate_limit(x, ratelimit)
+    end
+
+    return listenloop(h, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-"""
-Execute Request processing function `f(::HTTP.Request) -> HTTP.Response`.
-"""
-function handle_request(f::Function, http::Stream)
-    request::HTTP.Request = http.message
-    request.body = read(http)
-    request.response::HTTP.Response = f(request)
-    request.response.request = request
-    startwrite(http)
-    write(http, request.response.body)
-    return
+listen(f::Function, host, port::Integer; kw...) = listen(Handlers.Handler(f), host, port; kw...)
```

Comment:
Restricting callables to be subtypes of functions is bad, if it can be avoided.

Reply:
Fixed.